File: ios/Spandan/Sources/Signal/EnsembleAverageBeats.swift
import Foundation

/// Segments a pulse signal into individual cardiac cycles on the negative-
/// going zero crossing (steepest downslope, better temporal precision than
/// a peak-based split), duration-gates them, resamples each to a fixed
/// sample count via `PchipInterpolator`, time-warps each so its systolic
/// peak lands at a common cycle fraction (so beat-to-beat peak jitter
/// doesn't smear the notch across the average), keeps the top-correlation
/// fraction against a rough template, and coherently averages (trimmed mean
/// + median) -- the core deliverable Branch 2 exists for. Direct port of
/// android/app/.../signal/EnsembleAverageBeats.kt (read directly before
/// writing this), itself a port of
/// `matlab/src/morphology/ensembleAverageBeats.m`.
///
/// Returns nil on too few beats/survivors (the Android port throws
/// `IllegalStateException` here, matching the MATLAB source's own hard
/// `error(...)` calls exactly; this Swift port uses the Optional
/// convention instead -- `MorphologyWaveformEstimator`, the live caller, is
/// responsible for treating nil as a signal-quality problem, not a crash,
/// the same "ported function stays MATLAB-faithful, the live orchestrator
/// adds the fail-soft layer" split `RealHeartRateEstimator` already uses).
enum EnsembleAverageBeats {

    struct Options {
        var beatSamples = 256
        var systolicAnchorFraction = 0.25
        var durationRejectFraction = 0.30
        var qualityKeepFraction = 0.25
        var trimPercent = 20.0
    }

    struct Prototype {
        let trimmedMean: [Double]
        let median: [Double]
    }
    struct IqrBand {
        let q1: [Double]
        let q3: [Double]
        let width: [Double]
        let meanWidth: Double
        let meanWidthNormalized: Double
    }
    struct Stats {
        let beatsFound: Int
        let beatsRejectedByDuration: Int
        let beatsRejectedByQuality: Int
        let beatsAveraged: Int
        let snrGainEstimate: Double
    }
    struct Result {
        let prototype: Prototype
        let iqrBand: IqrBand
        let beatMatrix: [[Double]]
        let stats: Stats
    }

    static func apply(_ sig: [Double], fs: Double, opts: Options = Options()) -> Result? {
        let n = sig.count
        let timeAxis = (0..<n).map { Double($0) / fs }

        // --- Step 1: negative-going zero crossings of the zero-mean signal.
        let mean = sig.reduce(0.0, +) / Double(n)
        let zeroMean = sig.map { $0 - mean }

        var crossingTimes: [Double] = []
        for i in 0..<(n - 1) {
            if zeroMean[i] >= 0.0 && zeroMean[i + 1] < 0.0 {
                let frac = zeroMean[i] / (zeroMean[i] - zeroMean[i + 1])
                crossingTimes.append(timeAxis[i] + frac * (timeAxis[i + 1] - timeAxis[i]))
            }
        }

        let numBeatsFound = crossingTimes.count - 1
        guard numBeatsFound >= 3 else { return nil } // "need at least 3 to form an ensemble average"

        // --- Step 2: duration gate.
        let beatDurations = (0..<numBeatsFound).map { crossingTimes[$0 + 1] - crossingTimes[$0] }
        let medianDuration = median(beatDurations)

        let survivingBeatIdx = beatDurations.indices.filter { i in
            abs(beatDurations[i] - medianDuration) / medianDuration <= opts.durationRejectFraction
        }
        let beatsRejectedByDuration = numBeatsFound - survivingBeatIdx.count
        guard survivingBeatIdx.count >= 2 else { return nil } // "need at least 2 to form an ensemble average"

        // --- Step 3: resample each surviving beat to beatSamples via PCHIP.
        let beatSamples = opts.beatSamples
        let numSurviving = survivingBeatIdx.count
        var beatMatrixResampled = [[Double]](repeating: [Double](repeating: 0.0, count: beatSamples), count: numSurviving)

        for rowIdx in 0..<numSurviving {
            let k = survivingBeatIdx[rowIdx]
            let t0 = crossingTimes[k]
            let t1 = crossingTimes[k + 1]
            let queryTimes = (0..<beatSamples).map { i in t0 + (t1 - t0) * Double(i) / Double(beatSamples - 1) }
            beatMatrixResampled[rowIdx] = PchipInterpolator.interpolate(timeAxis, sig, queryTimes)
        }

        // --- Step 4: two-anchor time warp -- align the systolic peak to
        // systolicAnchorFraction of the cycle.
        let targetFrac = (0..<beatSamples).map { Double($0) / Double(beatSamples - 1) }
        let origFrac = targetFrac
        let peakAnchorFrac = opts.systolicAnchorFraction

        var beatMatrixWarped = [[Double]](repeating: [Double](repeating: 0.0, count: beatSamples), count: numSurviving)
        for rowIdx in 0..<numSurviving {
            let beatValues = beatMatrixResampled[rowIdx]
            var peakIdx = 0
            var peakVal = beatValues[0]
            for i in 1..<beatSamples where beatValues[i] > peakVal { peakVal = beatValues[i]; peakIdx = i }
            let peakFracOrig = origFrac[peakIdx]

            if peakFracOrig <= 0.0 || peakFracOrig >= 1.0 {
                // Degenerate: peak at the very first/last sample -- the
                // piecewise-linear warp below is undefined. Leave unwarped
                // rather than fabricate a mapping; the quality gate (Step 5)
                // is expected to down-weight it anyway.
                beatMatrixWarped[rowIdx] = beatValues
                continue
            }

            var newFrac = [Double](repeating: 0.0, count: beatSamples)
            for i in 0..<beatSamples {
                if origFrac[i] <= peakFracOrig {
                    newFrac[i] = origFrac[i] * (peakAnchorFrac / peakFracOrig)
                } else {
                    newFrac[i] = peakAnchorFrac + (origFrac[i] - peakFracOrig) * ((1.0 - peakAnchorFrac) / (1.0 - peakFracOrig))
                }
            }
            beatMatrixWarped[rowIdx] = PchipInterpolator.interpolate(newFrac, beatValues, targetFrac)
        }

        // --- Step 5: two-pass quality gate.
        var roughTemplate = [Double](repeating: 0.0, count: beatSamples)
        for col in 0..<beatSamples {
            var s = 0.0
            for rowIdx in 0..<numSurviving { s += beatMatrixWarped[rowIdx][col] }
            roughTemplate[col] = s / Double(numSurviving)
        }

        let correlations = (0..<numSurviving).map { pearsonCorr(beatMatrixWarped[$0], roughTemplate) }
        let sortOrder = (0..<numSurviving).sorted { correlations[$0] > correlations[$1] }
        let numKeep = max(1, Int((Double(numSurviving) * opts.qualityKeepFraction).rounded(.up)))
        let keepIdx = Array(sortOrder.prefix(numKeep))
        let beatsRejectedByQuality = numSurviving - numKeep

        let beatMatrix = keepIdx.map { beatMatrixWarped[$0] }

        // --- Step 6: trimmed mean + median.
        let trimmedMean = (0..<beatSamples).map { col in trimmedMeanOfColumn(beatMatrix, col: col, trimPercent: opts.trimPercent) }
        let medianRow = (0..<beatSamples).map { col in median(beatMatrix.map { $0[col] }) }

        // --- Step 7: per-sample IQR.
        var q1 = [Double](repeating: 0.0, count: beatSamples)
        var q3 = [Double](repeating: 0.0, count: beatSamples)
        for col in 0..<beatSamples {
            let columnData = beatMatrix.map { $0[col] }.sorted()
            q1[col] = percentile(columnData, 25.0)
            q3[col] = percentile(columnData, 75.0)
        }
        let width = (0..<beatSamples).map { q3[$0] - q1[$0] }
        let meanWidth = width.reduce(0.0, +) / Double(width.count)
        let prototypeRange = (trimmedMean.max() ?? 0.0) - (trimmedMean.min() ?? 0.0)
        let meanWidthNormalized = prototypeRange > 0.0 ? meanWidth / prototypeRange : Double.nan

        let stats = Stats(
            beatsFound: numBeatsFound,
            beatsRejectedByDuration: beatsRejectedByDuration,
            beatsRejectedByQuality: beatsRejectedByQuality,
            beatsAveraged: beatMatrix.count,
            snrGainEstimate: Double(beatMatrix.count).squareRoot()
        )

        return Result(
            prototype: Prototype(trimmedMean: trimmedMean, median: medianRow),
            iqrBand: IqrBand(q1: q1, q3: q3, width: width, meanWidth: meanWidth, meanWidthNormalized: meanWidthNormalized),
            beatMatrix: beatMatrix,
            stats: stats
        )
    }

    private static func median(_ x: [Double]) -> Double {
        let sorted = x.sorted()
        let n = sorted.count
        return n % 2 == 1 ? sorted[n / 2] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0
    }

    /// Linear-interpolation-between-order-statistics percentile, matching
    /// the MATLAB source's own manual `localPercentile` (MATLAB `prctile`'s
    /// default convention). `sortedAscending` must already be sorted.
    private static func percentile(_ sortedAscending: [Double], _ p: Double) -> Double {
        let n = sortedAscending.count
        if n == 1 { return sortedAscending[0] }
        let position = (p / 100.0) * Double(n - 1)
        let lower = Int(position.rounded(.down))
        let upper = Int(position.rounded(.up))
        let weight = position - Double(lower)
        return sortedAscending[lower] * (1.0 - weight) + sortedAscending[upper] * weight
    }

    /// Column-wise trimmed mean, matching MATLAB's `trimmean(X, percent, 1)`:
    /// removes `percent`% of values total (half from each tail) before averaging.
    private static func trimmedMeanOfColumn(_ matrix: [[Double]], col: Int, trimPercent: Double) -> Double {
        let values = matrix.map { $0[col] }.sorted()
        let n = values.count
        let trimEachSide = Int((Double(n) * (trimPercent / 100.0) / 2.0).rounded())
        if 2 * trimEachSide >= n { return values.reduce(0.0, +) / Double(n) } // degenerate: trimming everything -- fall back to plain mean
        var sum = 0.0
        var count = 0
        for i in trimEachSide..<(n - trimEachSide) { sum += values[i]; count += 1 }
        return sum / Double(count)
    }

    /// Manual Pearson correlation, matching the MATLAB source's own manual
    /// `pearsonCorrRow` (no Statistics Toolbox dependency there either).
    private static func pearsonCorr(_ x: [Double], _ y: [Double]) -> Double {
        let mx = x.reduce(0.0, +) / Double(x.count)
        let my = y.reduce(0.0, +) / Double(y.count)
        var num = 0.0
        var sxx = 0.0
        var syy = 0.0
        for i in x.indices {
            let dx = x[i] - mx
            let dy = y[i] - my
            num += dx * dy
            sxx += dx * dx
            syy += dy * dy
        }
        return num / (sxx * syy).squareRoot()
    }
}

File: android_segment30_mediapipe/app/src/main/java/com/spandan/app/signal/EnsembleAverageBeats.kt
package com.spandan.app.signal

import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Segment 19 (Branch 2 morphology port) -- real port of
 * `matlab/src/morphology/ensembleAverageBeats.m` (read directly from source
 * before writing this): segments a pulse signal into individual cardiac
 * cycles on the negative-going zero crossing (steepest downslope, better
 * temporal precision than a peak-based split), duration-gates them,
 * resamples each to a fixed sample count via [PchipInterpolator], time-warps
 * each so its systolic peak lands at a common cycle fraction (so beat-to-
 * beat peak jitter doesn't smear the notch across the average), keeps the
 * top-correlation fraction against a rough template, and coherently
 * averages (trimmed mean + median) -- the core deliverable this project's
 * whole Branch 2 exists for.
 *
 * @throws IllegalStateException on too few beats/survivors -- matching
 *   `ensembleAverageBeats.m`'s own hard `error(...)` calls exactly. This is
 *   a faithful, unsoftened port; [MorphologyWaveformEstimator] (the live
 *   caller) is responsible for catching this and failing soft (treating it
 *   as a signal-quality problem, not crashing), the same "ported function
 *   stays MATLAB-faithful, the live orchestrator adds the fail-soft layer"
 *   split [RealHeartRateEstimator] already uses around [BandpassFilter]/
 *   [HeartRateFft].
 */
object EnsembleAverageBeats {

    data class Options(
        val beatSamples: Int = 256,
        val systolicAnchorFraction: Double = 0.25,
        val durationRejectFraction: Double = 0.30,
        val qualityKeepFraction: Double = 0.25,
        val trimPercent: Double = 20.0
    )

    data class Prototype(val trimmedMean: DoubleArray, val median: DoubleArray)
    data class IqrBand(val q1: DoubleArray, val q3: DoubleArray, val width: DoubleArray, val meanWidth: Double, val meanWidthNormalized: Double)
    data class Stats(val beatsFound: Int, val beatsRejectedByDuration: Int, val beatsRejectedByQuality: Int, val beatsAveraged: Int, val snrGainEstimate: Double)
    data class Result(val prototype: Prototype, val iqrBand: IqrBand, val beatMatrix: Array<DoubleArray>, val stats: Stats)

    fun apply(sig: DoubleArray, fs: Double, opts: Options = Options()): Result {
        val n = sig.size
        val timeAxis = DoubleArray(n) { it / fs }

        // --- Step 1: negative-going zero crossings of the zero-mean signal.
        val mean = sig.average()
        val zeroMean = DoubleArray(n) { sig[it] - mean }

        val crossingTimes = mutableListOf<Double>()
        for (i in 0 until n - 1) {
            if (zeroMean[i] >= 0.0 && zeroMean[i + 1] < 0.0) {
                val frac = zeroMean[i] / (zeroMean[i] - zeroMean[i + 1])
                crossingTimes.add(timeAxis[i] + frac * (timeAxis[i + 1] - timeAxis[i]))
            }
        }

        val numBeatsFound = crossingTimes.size - 1
        check(numBeatsFound >= 3) {
            "Only ${maxOf(numBeatsFound, 0)} beat(s) found from negative-going zero crossings -- need at least 3 to form an ensemble average."
        }

        // --- Step 2: duration gate.
        val beatDurations = DoubleArray(numBeatsFound) { crossingTimes[it + 1] - crossingTimes[it] }
        val medianDuration = median(beatDurations)

        val survivingBeatIdx = beatDurations.indices.filter { i ->
            kotlin.math.abs(beatDurations[i] - medianDuration) / medianDuration <= opts.durationRejectFraction
        }
        val beatsRejectedByDuration = numBeatsFound - survivingBeatIdx.size
        check(survivingBeatIdx.size >= 2) {
            "Only ${survivingBeatIdx.size} beat(s) survived the duration gate -- need at least 2 to form an ensemble average."
        }

        // --- Step 3: resample each surviving beat to beatSamples via PCHIP.
        val beatSamples = opts.beatSamples
        val numSurviving = survivingBeatIdx.size
        val beatMatrixResampled = Array(numSurviving) { DoubleArray(beatSamples) }

        for (rowIdx in 0 until numSurviving) {
            val k = survivingBeatIdx[rowIdx]
            val t0 = crossingTimes[k]
            val t1 = crossingTimes[k + 1]
            val queryTimes = DoubleArray(beatSamples) { i -> t0 + (t1 - t0) * i / (beatSamples - 1) }
            beatMatrixResampled[rowIdx] = PchipInterpolator.interpolate(timeAxis, sig, queryTimes)
        }

        // --- Step 4: two-anchor time warp -- align the systolic peak to
        // systolicAnchorFraction of the cycle.
        val targetFrac = DoubleArray(beatSamples) { it.toDouble() / (beatSamples - 1) }
        val origFrac = targetFrac
        val peakAnchorFrac = opts.systolicAnchorFraction

        val beatMatrixWarped = Array(numSurviving) { DoubleArray(beatSamples) }
        for (rowIdx in 0 until numSurviving) {
            val beatValues = beatMatrixResampled[rowIdx]
            var peakIdx = 0
            var peakVal = beatValues[0]
            for (i in 1 until beatSamples) if (beatValues[i] > peakVal) { peakVal = beatValues[i]; peakIdx = i }
            val peakFracOrig = origFrac[peakIdx]

            if (peakFracOrig <= 0.0 || peakFracOrig >= 1.0) {
                // Degenerate: peak at the very first/last sample -- the
                // piecewise-linear warp below is undefined. Leave unwarped
                // rather than fabricate a mapping; the quality gate (Step 5)
                // is expected to down-weight it anyway.
                beatMatrixWarped[rowIdx] = beatValues.copyOf()
                continue
            }

            val newFrac = DoubleArray(beatSamples)
            for (i in 0 until beatSamples) {
                newFrac[i] = if (origFrac[i] <= peakFracOrig) {
                    origFrac[i] * (peakAnchorFrac / peakFracOrig)
                } else {
                    peakAnchorFrac + (origFrac[i] - peakFracOrig) * ((1.0 - peakAnchorFrac) / (1.0 - peakFracOrig))
                }
            }
            beatMatrixWarped[rowIdx] = PchipInterpolator.interpolate(newFrac, beatValues, targetFrac)
        }

        // --- Step 5: two-pass quality gate.
        val roughTemplate = DoubleArray(beatSamples) { col ->
            var s = 0.0
            for (rowIdx in 0 until numSurviving) s += beatMatrixWarped[rowIdx][col]
            s / numSurviving
        }

        val correlations = DoubleArray(numSurviving) { pearsonCorr(beatMatrixWarped[it], roughTemplate) }
        val sortOrder = (0 until numSurviving).sortedByDescending { correlations[it] }
        val numKeep = maxOf(1, ceil(numSurviving * opts.qualityKeepFraction).toInt())
        val keepIdx = sortOrder.take(numKeep)
        val beatsRejectedByQuality = numSurviving - numKeep

        val beatMatrix = Array(numKeep) { beatMatrixWarped[keepIdx[it]] }

        // --- Step 6: trimmed mean + median.
        val trimmedMean = DoubleArray(beatSamples) { col -> trimmedMeanOfColumn(beatMatrix, col, opts.trimPercent) }
        val medianRow = DoubleArray(beatSamples) { col -> median(DoubleArray(beatMatrix.size) { r -> beatMatrix[r][col] }) }

        // --- Step 7: per-sample IQR.
        val q1 = DoubleArray(beatSamples)
        val q3 = DoubleArray(beatSamples)
        for (col in 0 until beatSamples) {
            val columnData = DoubleArray(beatMatrix.size) { r -> beatMatrix[r][col] }.also { it.sort() }
            q1[col] = percentile(columnData, 25.0)
            q3[col] = percentile(columnData, 75.0)
        }
        val width = DoubleArray(beatSamples) { q3[it] - q1[it] }
        val meanWidth = width.average()
        val prototypeRange = (trimmedMean.max() - trimmedMean.min())
        val meanWidthNormalized = if (prototypeRange > 0.0) meanWidth / prototypeRange else Double.NaN

        val stats = Stats(
            beatsFound = numBeatsFound,
            beatsRejectedByDuration = beatsRejectedByDuration,
            beatsRejectedByQuality = beatsRejectedByQuality,
            beatsAveraged = beatMatrix.size,
            snrGainEstimate = sqrt(beatMatrix.size.toDouble())
        )

        return Result(
            prototype = Prototype(trimmedMean, medianRow),
            iqrBand = IqrBand(q1, q3, width, meanWidth, meanWidthNormalized),
            beatMatrix = beatMatrix,
            stats = stats
        )
    }

    private fun median(x: DoubleArray): Double {
        val sorted = x.copyOf().also { it.sort() }
        val n = sorted.size
        return if (n % 2 == 1) sorted[n / 2] else (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0
    }

    /** Linear-interpolation-between-order-statistics percentile, matching
     *  `ensembleAverageBeats.m`'s own manual `localPercentile` (MATLAB
     *  `prctile`'s default convention). [sortedAscending] must already be sorted. */
    private fun percentile(sortedAscending: DoubleArray, p: Double): Double {
        val n = sortedAscending.size
        if (n == 1) return sortedAscending[0]
        val position = (p / 100.0) * (n - 1)
        val lower = floor(position).toInt()
        val upper = ceil(position).toInt()
        val weight = position - lower
        return sortedAscending[lower] * (1.0 - weight) + sortedAscending[upper] * weight
    }

    /** Column-wise trimmed mean, matching MATLAB's `trimmean(X, percent, 1)`:
     *  removes `percent`% of values total (half from each tail) before averaging. */
    private fun trimmedMeanOfColumn(matrix: Array<DoubleArray>, col: Int, trimPercent: Double): Double {
        val values = DoubleArray(matrix.size) { matrix[it][col] }.also { it.sort() }
        val n = values.size
        val trimEachSide = Math.round(n * (trimPercent / 100.0) / 2.0).toInt()
        if (2 * trimEachSide >= n) return values.average() // degenerate: trimming everything -- fall back to plain mean
        var sum = 0.0
        var count = 0
        for (i in trimEachSide until n - trimEachSide) { sum += values[i]; count++ }
        return sum / count
    }

    /** Manual Pearson correlation, matching `ensembleAverageBeats.m`'s own
     *  manual `pearsonCorrRow` (no Statistics Toolbox dependency there either). */
    private fun pearsonCorr(x: DoubleArray, y: DoubleArray): Double {
        val mx = x.average()
        val my = y.average()
        var num = 0.0
        var sxx = 0.0
        var syy = 0.0
        for (i in x.indices) {
            val dx = x[i] - mx
            val dy = y[i] - my
            num += dx * dy
            sxx += dx * dx
            syy += dy * dy
        }
        return num / sqrt(sxx * syy)
    }
}
